snuba/datasets/entity.py

```diff
@@ -156,3 +175,106 @@ def describe(self) -> Description:
+    disallowed_aggregations: Optional[Sequence[str]] = None
+    organization: Optional[int] = None
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        """
+        Applies entity specific validations on query argument passed
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> EntitySubscription:
+        raise NotImplementedError
+
+
+class EntitySubscriptionValidation(EntitySubscription):
+    def __init__(
+        self,
+        max_allowed_aggregations: int = 1,
+        disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"],
+    ) -> None:
+        self.max_allowed_aggregations: int = max_allowed_aggregations
+        self.disallowed_aggregations: Sequence[str] = disallowed_aggregations
+
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        # Import get_entity() when used, not at import time to avoid circular imports
+        from snuba.datasets.entities.factory import get_entity
+
+        # TODO: Support composite queries with multiple entities.
+        from_clause = query.get_from_clause()
+        if not isinstance(from_clause, EntityDS):
+            raise InvalidSubscriptionError("Only simple queries are supported")
+        entity = get_entity(from_clause.key)
+
+        SubscriptionAllowedClausesValidator(
+            self.max_allowed_aggregations, self.disallowed_aggregations
+        ).validate(query)
+        if entity.required_time_column:
+            NoTimeBasedConditionValidator(entity.required_time_column).validate(query)
+
+
+class BaseEntitySubscription(EntitySubscriptionValidation):
+    def __init__(
+        self,
+        max_allowed_aggregations: Optional[int] = None,
+        disallowed_aggregations: Optional[Sequence[str]] = None,
+        data_dict: Mapping[str, Any] = {},
+    ) -> None:
+        super().__init__()
+        if max_allowed_aggregations and disallowed_aggregations:
+            super().__init__(max_allowed_aggregations, disallowed_aggregations)
+        self.set_org(data_dict)
+
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> BaseEntitySubscription:
+        if "organization" in data_dict:
+            try:
+                self.organization: int = data_dict["organization"]
+            except KeyError:
+                raise InvalidQueryException(
+                    "organization param is required for any query over sessions entity"
```

Comment:
This is kind of weird. why are we talking about a specific product entity in the error message? I think this whole class could use a docstring which describes:

1. the function it accomplishes and how it fits into the larger system
2. usage examples not tied to any specific entity

Reply:
Right, this was copy pasta from the original EntitySubscription (which all extended from SessionsSubscription). Will update this. Sounds good, will add a descriptive docstring with examples.

Reply:
Moved files back to original directories as suggested.